### ISSUE TITLE
pre mode with "latest" dist-tag publishing?

### ISSUE THREAD

**Issue by FredKSchott** [changesets/changesets#786]
Hey everyone! Fred from Astro here, and we've just entered a 2-month `v1.0.0-beta.X` period before our official Astro v1.0 release in June.

We've been happy changeset users so far but now I'm not sure the best way to structure/configure changesets for the next two months. What I'd really want is...

### What we want:

- Every package in the monorepo released to `latest`
- Only `astro` is versioned at `v1.0.0-beta.X`, all others follow v0.x semver
- don't need to be in `pre` mode if we can help it.

### What we've tried:

But I don't think that's possible. Based on what I know of changesets, I expected this to work as a secondary option:

- monorepo is in `pre` mode (tag: 'beta')
- when we publish, we can still publish all packages to `latest`

The problem is that `changeset publish --tag latest` doesn't seem to be respected. When I run a dry-run, I see an error that `--tag` isn't supported in `pre` mode.

Any thoughts for how we could achieve something like this with changesets? Since it's short term, I'm willing to do something hacky if you have any suggestions. Thank you!

**Comment by Andarist** [comment]
What should be the tag for the astro? Should it be `beta` here? Should the `astro@latest` keep pointing to some older `0.x` version?

> The problem is that changeset publish --tag latest doesn't seem to be respected. When I run a dry-run, I see an error that --tag isn't supported in pre mode.

Even if this would work... what would happen to `astro`? `changeset publish` tries to publish *everything* that has not yet been published to the registry. And it seems that you want to avoid publishing `astro` beta versions as latest (unless I misunderstood the issue).

If this would help you - I'm open to scheduling a quick call to hash it out cause I think there is definitely some workaround (even if slightly hacky :P) to make it work for you. I just need to understand all of your constraints/requirements properly

**Comment by diegohaz** [comment]
From what I understand, they want to publish astro@1.0.0-beta.X as the `latest` tag on npm. I have the same requirement with [`ariakit`](https://www.npmjs.com/package/ariakit).

**Comment by Andarist** [comment]
> Update 2: I now understand that it was released on the latest tag because I didn't have any stable release on the ariakit-changesets-test package.

Yes, the `latest` tag has to be set to something - that's a requirement of the npm registry.

> Update 3: I was able to force the latest tag on the prerelease by using the publishConfig field on the package.json file.

Interesting. So you have just added something like this?
```json
{
  "publishConfig": { "tag": "latest" }
}
```

Which npm client do you use?

**Comment by FredKSchott** [comment]
> Update 3: I was able to force the `latest` tag on the prerelease by using the `publishConfig` field on the package.json file.

Interesting! @diegohaz  Did this work automatically with `changesets publish`, or did you have to define your own publish step using pnpm/yarn/npm, outside of changesets?

**Comment by Andarist** [comment]
I'm pretty sure that this is not handled by Changesets and that it might only be handled by *some* package managers (I'm unsure which ones support this).

**Comment by FredKSchott** [comment]
> If this would help you - I'm open to scheduling a quick call to hash it out cause I think there is definitely some workaround (even if slightly hacky :P) to make it work for you. I just need to understand all of your constraints/requirements properly

@Andarist apologies for missing this! We're in the middle of a big launch week which has been quite busy 😅 . If you're interested in joining us in Discord, we could do this async but maybe a bit more high-touch? astro.build/chat

Otherwise, happy to keep debugging here!

**Comment by diegohaz** [comment]
I'm using npm 8.

**Comment by FredKSchott** [comment]
@Andarist honestly, I'm starting to think we might want to fork changesets, and do an astro-specific hack to get us through the next 2 months. I keep getting stuck on ideas:

- Run a post-publish `npm dist-tag add astro@1.0.0-beta.X latest` step to our GitHub Action, so that `latest` always matches beta (okay, but we need this for all packages in the monorepo, not just Astro)
- Run a post-publish `npm dist-tag add X@Y latest` step to our GitHub Action, so that `latest` always matches beta for all packages in the monorepo (okay, but non-Astro dependencies are getting the `-beta.1` unnecessarily. We kind of want them to be non-beta, non-pre, normal semver)
- I can't think of any way this could work without the repo being in pre-mode.

I think the smallest fork we could make, would be to keep the monorepo in normal mode, but then surgically overwrite whatever internal logic handles version bumping, to just always handle `"astro"` as a special case to always bump as `1.0.0-beta${X+1}`. Everything else about normal "non-pre" mode would remain the same.

@Andarist what do you think?

**Comment by Andarist** [comment]
Yeah, forking seems to be the easiest solution to this - especially given that you only need this temporarily~.

In general, Changesets has not been designed to handle multiple independent "release plans". It's not impossible to add support for those... but it raises the complexity quite a bit because we would have to consider how parallel release plans interact with each other, can they overlap somehow? Can A from RP1 depend on B from RP2? etc etc. This would require deeper thought, analysis, and implementation - and, unfortunately, I'm so slammed with piling OSS issues that I just don't have capacity to do all of this on my own.

**Comment by dominikg** [comment]
> * Run a post-publish `npm dist-tag add astro@1.0.0-beta.X latest` step to our GitHub Action, so that `latest` always matches beta (okay, but we need this for all packages in the monorepo, not just Astro)

about that: https://github.com/sveltejs/kit/pull/4564

**Comment by matthewp** [comment]
@Andarist 

> What should be the tag for the astro? Should it be beta here? Should the astro@latest keep pointing to some older 0.x version?

No, the tag should be `latest`.

This doesn't seem to go against what changeset is doing that much to me. We just want it to be the `latest` tag instead of the beta tag.

---

@FredKSchott 

> Run a post-publish npm dist-tag add X@Y latest step to our GitHub Action, so that latest always matches beta for all packages in the monorepo (okay, but non-Astro dependencies are getting the -beta.1 unnecessarily. We kind of want them to be non-beta, non-pre, normal semver)

Do we? Why? It seems perfectly fine for everything to be marked with the beta version but still tagged latest.

I agree with @Andarist here that this is the part that seems undoable with how changeset works, you're either in pre-mode or not. I think being in pre-mode for everything is fine, but just with the difference of the tag being `latest` rather than the pre tag (`beta`, in this case).

**Comment by Andarist** [comment]
> This doesn't seem to go against what changeset is doing that much to me. We just want it to be the latest tag instead of the beta tag.

Maybe we could actually respect `publishConfig.tag` if that's a thing supported by other tools - I would love to know which tools actually support this.

This implies though that all your versions would be "preversions" - the only difference would be in the published tag.

**Comment by diegohaz** [comment]
> Maybe we could actually respect publishConfig.tag if that's a thing supported by other tools - I would love to know which tools actually support this.


From what I tested, Changesets already supports this (maybe npm just forces this tag and ignores whatever tag you pass to the npm publish command).

As for other tools, I've used the same approach with lerna in the past.
